### ISSUE TITLE
Song callback event doesn't happen correctly when using radio streams.

### ISSUE THREAD

**Issue by HuggableSquare** [archSeer/ruby-mpd#20]
When using a song callback as such:
```
@mpd.on :song do |current|
  puts current.inspect
end
```
it only works when a new file starts to play, which is a problem with radio streams, for two reasons.

* It takes a second or two to get the metadata from the stream, and so when you first add the stream the callback happens, but it generates something like this:
```
#<MPD::Song:0x00000002a30c58 @data={:pos=>0, :id=>703}, @time=nil, @file="http://streamer.strobe.fm/", @title=nil, @artist=nil, @album=nil, @albumartist=nil>
```
But, it would generate something like this if the callback detected the metadata changing:
```
#<MPD::Song:0x000000019a9628 @data={:name=>"Strobe.FM", :pos=>0, :id=>703}, @time=nil, @file="http://streamer.strobe.fm/", @title="Goo Goo Dolls - Caught In The Storm", @artist=nil, @album=nil, @albumartist=nil>
```

* Whenever the song changes in the stream, it doesn't go through the code again, and so the metadata that was generated when the stream was first added stays that way until it's changed to a different file.

The only thing I can think of is that the callback checks if the id changed, (notice the id is 703 in both of those arrays, because it's still the same stream) but it doesn't change when using radio streams. I guess the best way to check this would be to check if anything in the array changed.

Obviously my diagnosis could be entirely wrong, I'm not that good with ruby, but that's my theory.

**Comment by AStrangeEnigma** [comment]
I came here to leave the exact same issue, and ended up running into this! I didn't think someone else would have the exact same problem. I would really love if this could be supported. I was thinking that a potential solution would be adding another callback for the metadata - so whenever Title changes, it runs the callback. Not sure how viable that is.

Huggable, thanks for posting a workaround.

**Comment by archSeer** [comment]
@whomwah refactored the callback processor, which might fix this, but we never really got it merged. https://github.com/archSeer/ruby-mpd/pull/32

**Comment by AStrangeEnigma** [comment]
Thanks for the rapid response @archSeer ! My own workaround was to add a title field to status like so:

brick@treef ~/mumblecop/ruby-mpd (git)-[master] % git diff
diff --git a/lib/ruby-mpd.rb b/lib/ruby-mpd.rb
index d207a9d..850808c 100644
--- a/lib/ruby-mpd.rb
+++ b/lib/ruby-mpd.rb
@@ -222,6 +222,7 @@ private
         status[:time] ||= [nil, nil] # elapsed, total
         status[:audio] ||= [nil, nil, nil] # samp, bits, chans
         status[:song] = mpd.current_song rescue nil
+        status[:title] = mpd.current_song.title rescue nil
         status[:updating_db] ||= nil
 
         status.each do |key, val|


But I'm glad to hear that there is something in the works that will fix it in a release :) My own solution definitely seems like a hack.

If you need anything from me to help you guys out, let me know.

**Comment by AStrangeEnigma** [comment]
I did some digging and I think I found a rather smooth solution to this issue! The thing is, the MPD song details change when the radio track changes but comparing the songs with == returns true - even though the object has changed! Consider this:

```
head :029 > test
 => #<MPD::Song:0x000000016c9750 @data={:name=>"RPGN Radio", :pos=>0, :id=>51}, @time=nil, @file="http://stream.rpgamers.net:8000/rpgn", @title="Infinite Undiscovery - Pure Alabaster", @artist=nil, @album=nil, @albumartist=nil>
head :030 > test2
 => #<MPD::Song:0x0000000168e1c8 @data={:name=>"RPGN Radio", :pos=>0, :id=>51}, @time=nil, @file="http://stream.rpgamers.net:8000/rpgn", @title="Okami - Shinshuu Plains I & II", @artist=nil, @album=nil, @albumartist=nil>
head :033 > test === test2
 => true 
head :034 > test == test2
 => true 
head :035 > test.eql? test2
 => false 
head :036 > test.equal? test2
 => false 
```

So, if we used .eql? or .equal? (I am not exactly sure which one is better) instead of == at https://github.com/archSeer/ruby-mpd/blob/master/lib/ruby-mpd.rb#L228 then the MPD song callback would trigger when the track in a stream changes. I think this makes this change makes a lot of sense. My only concern is that it might cause unexpected behavior with other callbacks (although I doubt it) but we should test and make sure everything still works as expected.

Also, I checked the refactored version in #32 and it seems that it would need to be changed in the same way.

Thanks, let me know what you think!

**Comment by AStrangeEnigma** [comment]
Created pull request #33 to discuss further

**Comment by archSeer** [comment]
Closed via #33.
